src/Toml/Parser/Value.hs

```diff
@@ -148,22 +146,19 @@ tableNameP = between (text "[") (text "]") keyP
 -- Values
 
 decimalP :: Parser Integer
-decimalP = mkInteger <$> decimalStringP
+decimalP = zero <|> more
   where
-    decimalStringP   = fst <$> match (some digitChar >> many _digitsP)
-    _digitsP         = try (char '_') >> some digitChar
-    mkInteger        = textToInt . stripUnderscores
-    textToInt        = fst . fromRight (error "Underscore parser has a bug") . TR.decimal
-    stripUnderscores = Text.filter (/= '_')
-
+    zero = 0 <$ char '0'
+    more = fst . head . readDec . concat <$> sepBy1 (some digitChar) (char '_')
 
 integerP :: Parser Integer
-integerP = lexeme (bin <|> oct <|> hex <|> dec) <?> "integer"
+integerP = lexeme ((bin <|> oct <|> hex <|> dec) <* lookAhead sep) <?> "integer"
```

Comment:
Hmm, this change requires some thinking 🤔 🤔 🤔 

I also wonder how it affects performance. Let's wait for the benchmarks to be merged to `master` so we can compare performance of this function before and after this change.

Reply:
OK.